packages/tailwindcss/src/utilities.ts

```diff
@@ -4569,3 +4584,360 @@ export function createUtilities(theme: Theme) {
+                usedValueFn ||= false
+                replaceDeclarationWith([])
+                return ValueParser.ValueWalkAction.Stop
+              }
+
+              // Modifier function, e.g.: `--modifier(integer)`
+              else if (valueNode.value === '--modifier') {
+                // If there is no modifier present in the candidate, then the
+                // declaration can be removed.
+                if (modifier === null) {
+                  replaceDeclarationWith([])
+                  return ValueParser.ValueWalkAction.Skip
+                }
+
+                usedModifierFn = true
+
+                let replacement = resolveValueFunction(modifier, valueNode, designSystem)
+                if (replacement) {
+                  resolvedModifierFn = true
+                  replaceWith(replacement.nodes)
+                  return ValueParser.ValueWalkAction.Skip
+                }
+
+                // Drop the declaration in case we couldn't resolve the value
+                usedModifierFn ||= false
+                replaceDeclarationWith([])
+                return ValueParser.ValueWalkAction.Stop
+              }
+            }) ?? ValueParser.ValueWalkAction.Continue
+
+          if (result === ValueParser.ValueWalkAction.Continue) {
+            node.value = ValueParser.toCss(valueAst)
+          }
+        })
+
+        // Used `--value(…)` but nothing resolved
+        if (usedValueFn && !resolvedValueFn) return null
+
+        // Used `--modifier(…)` but nothing resolved
+        if (usedModifierFn && !resolvedModifierFn) return null
+
+        // Resolved `--value(ratio)` and `--modifier(…)`, which is invalid
+        if (resolvedRatioValue && resolvedModifierFn) return null
+
+        // When a candidate has a modifier, then the `--modifier(…)` must
+        // resolve correctly or the `--value(ratio)` must resolve correctly.
+        if (modifier && !resolvedRatioValue && !resolvedModifierFn) return null
+
+        // Resolved `--value(ratio)`, so all other declarations that didn't use
+        // `--value(ratio)` should be removed. E.g.: `--value(number)` would
+        // otherwise resolve for `foo-1/2`.
+        if (resolvedRatioValue) {
+          for (let [declaration, parent] of resolvedDeclarations) {
+            let idx = parent.nodes.indexOf(declaration)
+            if (idx !== -1) parent.nodes.splice(idx, 1)
+          }
+        }
+
+        return atRule.nodes
+      })
+
+      designSystem.utilities.suggest(name.slice(0, -2), () => {
+        return [
+          {
+            values: designSystem.theme
+              .keysInNamespaces(valueThemeKeys)
+              .map((x) => x.replaceAll('_', '.')),
+            modifiers: designSystem.theme
+              .keysInNamespaces(modifierThemeKeys)
+              .map((x) => x.replaceAll('_', '.')),
+          },
+        ] satisfies SuggestionGroup[]
+      })
+    }
+  }
+
+  if (IS_VALID_UTILITY_NAME.test(name)) {
```

Comment:
```suggestion
  if (IS_VALID_STATIC_UTILITY_NAME.test(name)) {
```

should we rename this too now for consistency with the previous `IS_VALID_FUNCTIONAL_UTILITY_NAME` branch?